lib/Relation.php

```diff
@@ -0,0 +1,557 @@
+     * ])
+     * # SELECT * FROM users WHERE name = 'Joe' AND email = 'joe@example.com';
+     *
+     * # hash
+     * `where` will also accept a hash condition, in which the keys are fields
+     * and the values are values to be searched for.
+     *
+     * Fields can be symbols or strings. Values can be single values, arrays, or ranges.
+     *
+     * User.where([
+     *   'name' => "Joe",
+     *   'email' => "joe@example.com"
+     * ])
+     * # SELECT * FROM users WHERE name = 'Joe' AND email = 'joe@example.com'
+     *
+     * @return Relation<TModel>
+     */
+    public function where(): Relation
+    {
+        $this->options['conditions'] ??= [];
+
+        $args = func_get_args();
+        $numArgs = count($args);
+
+        if (1 != $numArgs) {
+            throw new \ArgumentCountError('`where` requires exactly one argument.');
+        }
+        $arg = $args[0];
+
+        $expression = WhereClause::from_arg($arg);
+
+        $this->options['conditions'][] = $expression;
+
+        return $this;
+    }
+
+    /**
+     * Sets readonly attributes for the returned relation. If value is true (default),
+     * attempting to update a record will result in an error.
+     *
+     * $users = User::readonly()
+     * $users->first()->save()  // throws exception ActiveRecord::ReadOnlyRecord: User is marked as readonly
+     */
+    public function readonly(bool $readonly): Relation
+    {
+        $this->options['readonly'] = $readonly;
+
+        return $this;
+    }
+
+    /**
+     * Find by id - This can either be a specific id (1),
+     * a list of ids (1, 5, 6), or an array of ids ([5, 6, 10]).
+     *
+     * If one or more records cannot be found for the requested ids,
+     * then ActiveRecord::RecordNotFound will be raised.
+     *
+     * If the primary key is an integer, find by id coerces its arguments by using to_i.
+     *
+     * Person.find(1)          # returns the object for ID = 1
+     * Person.find("1")        # returns the object for ID = 1
+     * Person.find("31-sarah") # returns the object for ID = 31
+     * Person.find(1, 2, 6)    # returns an array for objects with IDs in (1, 2, 6)
+     * Person.find([7, 17])    # returns an array for objects with IDs in (7, 17)
+     * Person.find([1])        # returns an array for the object with ID = 1
+     *
+     * Person.where("administrator = 1").order("created_on DESC").find(1)
+     *
+     * Person.find(-1)          # throws RecordNotFound
+     *
+     * Person.find()           # throws ValidationsArgumentError as there's no where statement to execute
+     *
+     * @throws RecordNotFound if any of the records cannot be found
+     *
+     * @return Model|array<Model>
+     */
+    public function find()
+    {
+        $args = func_get_args();
+        $num_args = count($args);
+        $class = $this->className;
+
+        if (0 === $num_args) {
+            throw new ValidationsArgumentError('find requires at least one argument');
+        }
+
```

Comment:
throw an exception if having clause is specified but group by is not.

Reply:
I'm gonna take your worf on this one (and I saw it in the docs), but I haven't really experimented with it.

3b29623